src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt

```diff
@@ -0,0 +1,369 @@
+                    .toMap(mutableMapOf())
+                or = queriedFields.remove("OR")?.second?.value?.let {
+                    return@let (it as ArrayValue).values
+                            ?: throw IllegalArgumentException("OR is expected to be a list")
+                }
+                and = queriedFields.remove("AND")?.second?.value?.let {
+                    return@let (it as ArrayValue).values
+                            ?: throw IllegalArgumentException("AND is expected to be a list")
+                }
+                for (definedField in type.fieldDefinitions) {
+                    if (definedField.isRelationship()) {
+                        collectRelationPredicates(type, definedField, queriedFields)
+                    } else {
+                        collectFieldPredicates(definedField, queriedFields)
+                    }
+                }
+                if (queriedFields.isNotEmpty()) {
+                    throw OptimizedQueryException("queried unknown fields: " + queriedFields.keys)
+                }
+                fieldPredicates.sortedBy(Predicate<FieldOperator>::index)
+                relationPredicates.sortedBy(Predicate<RelationOperator>::index)
+            }
+
+            private fun collectRelationPredicates(type: GraphQLFieldsContainer, definedField: GraphQLFieldDefinition, queriedFields: MutableMap<String, Pair<Int, ObjectField>>) {
+                RelationOperator.values()
+                    .map { it to definedField.name + it.suffix }
+                    .mapNotNull { (queryOp, queryFieldName) ->
+                        val (index, objectField) = queriedFields.remove(queryFieldName) ?: return@mapNotNull null
+                        val op = when (definedField.type.isList()) {
+                            true -> when (queryOp) {
+                                RelationOperator.NOT -> when (objectField.value) {
+                                    is NullValue -> RelationOperator.NOT
+                                    else -> RelationOperator.NONE
+                                }
+                                RelationOperator.EQ_OR_NOT_EXISTS -> when (objectField.value) {
+                                    is NullValue -> RelationOperator.EQ_OR_NOT_EXISTS
+                                    else -> {
+                                        LOGGER.info("$queryFieldName on type ${type.name} was used for filtering, consider using ${definedField.name}${RelationOperator.EVERY.suffix} instead")
+                                        RelationOperator.EVERY
+                                    }
+                                }
+                                else -> queryOp
+                            }
+                            false -> when (queryOp) {
+                                RelationOperator.SINGLE -> {
+                                    LOGGER.warn("Using $queryFieldName on type ${type.name} is deprecated, use ${definedField.name} directly")
```

Comment:
this would be really spammy in the logs, as it shows up every time, consider `debug` instead

Reply:
So the user will only notice after he enable debug log for this library

Reply:
actually what we should do is to provide the info as part of the translation response so that the middleware server can send it to the client. the server logs won't be read by the folks sending the queries anyway :)

Reply:
:+1:

Reply:
I created issue https://github.com/neo4j-graphql/neo4j-graphql-java/issues/80 for this